# Метод Монте-Карло
Оганджанян Артем, M3339, вариант 11

In [1]:
import numpy as np
import scipy.stats
import math

In [2]:
gamma = 0.95
t = scipy.stats.norm.ppf((1 + gamma) / 2)
print(t)

1.95996398454


$$t = \text{квантиль}\left(\frac{1+\gamma}{2}\right) = 1.9599639845$$

$$\Delta = \frac{t \cdot \sigma}{\sqrt{n}}$$

## Задание 1

$$f(x) = e^{-ax} \qquad k = 10 \qquad c = 8.8 \qquad a = 0.35$$

In [3]:
a = 0.35
f = lambda x : np.exp(-a * x)
k = 10
c = 8.8

In [4]:
def task_1(n):
    fx = f(np.random.uniform(size = n * k)).reshape((n, k))
    Fx = (np.apply_along_axis(np.sum, 1, fx) <= c).astype(np.float)
    mean = np.average(Fx)
    variance = np.average((Fx - mean) ** 2)
    deviation = math.sqrt(variance)
    confidence = t * deviation / math.sqrt(n)
    return (mean, confidence)

In [5]:
for n in [10000, 1000000]:
    value, confidence = task_1(n)
    print("{} итераций:\n\tОбъём {} ± {};\n\tДоверительный интервал ({}, {})"
        .format(n, value, confidence, value - confidence, value + confidence))
    print()

10000 итераций:
	Объём 0.906 ± 0.005719737747533486;
	Доверительный интервал (0.9002802622524665, 0.9117197377475336)

1000000 итераций:
	Объём 0.908992 ± 0.0005637257936707053;
	Доверительный интервал (0.9084282742063293, 0.9095557257936707)



## Задание 2

$$\int_0^\infty \sqrt{1+x^2} \exp(-3x) \mathrm{d}x \approx 0.364129 \text{ (Вольфрам)}$$

In [6]:
def task_2(n):
    fx = np.sqrt(1 + np.random.exponential(scale = 1/3, size = n) ** 2)
    mean = np.average(fx)
    variance = np.average((fx - mean) ** 2)
    deviation = math.sqrt(variance)
    confidence = t * deviation / math.sqrt(n)
    return (mean / 3, confidence / 3)

In [7]:
for n in [10000, 1000000]:
    value, confidence = task_2(n)
    print("{} итераций:\n\tИнтеграл {} ± {};\n\tДоверительный интервал ({}, {})"
        .format(n, value, confidence, value - confidence, value + confidence))
    print()

10000 итераций:
	Интеграл 0.3644574603454746 ± 0.0011294115056231994;
	Доверительный интервал (0.3633280488398514, 0.3655868718510978)

1000000 итераций:
	Интеграл 0.3641025784710807 ± 0.00011112799664332449;
	Доверительный интервал (0.3639914504744374, 0.364213706467724)



## Задание 3

$$\int_4^9 \frac{\ln(x)}{x+1} \mathrm{d}x \approx 1.24742 \text{ (Вольфрам)}$$

In [8]:
def task_3(n):
    xs = np.random.uniform(size = n, low = 4, high = 9)
    fx = np.log(xs) / (xs + 1)
    mean = np.average(fx)
    variance = np.average((fx - mean) ** 2)
    deviation = math.sqrt(variance)
    confidence = t * deviation / math.sqrt(n)
    return (mean * 5, confidence * 5)

In [9]:
for n in [10000, 1000000]:
    value, confidence = task_3(n)
    print("{} итераций:\n\tИнтеграл {} ± {};\n\tДоверительный интервал ({}, {})"
        .format(n, value, confidence, value - confidence, value + confidence))
    print()

10000 итераций:
	Интеграл 1.2466394781073546 ± 0.0017100330858498913;
	Доверительный интервал (1.2449294450215047, 1.2483495111932046)

1000000 итераций:
	Интеграл 1.247442748103257 ± 0.00017091110664507332;
	Доверительный интервал (1.247271836996612, 1.247613659209902)

